File: app/src/main/java/com/devidea/timeleft/AppWidget.java
package com.devidea.timeleft;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.os.SystemClock;
import android.util.Log;
import android.widget.RemoteViews;

import androidx.room.Room;

import static com.devidea.timeleft.MainActivity.appDatabase;
import static com.devidea.timeleft.MainActivity.itemGenerator;
import static com.devidea.timeleft.MainActivity.DEFAULT_TIME;
import static com.devidea.timeleft.MainActivity.DEFAULT_DAY;
import static com.devidea.timeleft.MainActivity.DEFAULT_YEAR;

public class AppWidget extends AppWidgetProvider {

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        String action = intent.getAction();
        int[] appWidgetIds;
        Log.d("widget", "onReceive() action = " + action);

        if (AppWidgetManager.ACTION_APPWIDGET_UPDATE.equals(action)) {
            if (appDatabase == null) {
                appDatabase = Room.databaseBuilder(context, AppDatabase.class, "ItemData")
                        .allowMainThreadQueries()
                        .build();
            }

            appWidgetIds = appDatabase.DatabaseDao().get();

            if (appWidgetIds != null && appWidgetIds.length > 0) {
                this.onUpdate(context, AppWidgetManager.getInstance(context), appWidgetIds);
            }
        }
    }


    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);
        // There may be multiple widgets active, so update all of them
        for (int appWidgetId : appWidgetIds) {
            Log.d("widget", "appWidgetId is " + appWidgetId);
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {
        Intent intent = new Intent(context, AppWidget.class);

        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        alarmManager.setInexactRepeating(AlarmManager.ELAPSED_REALTIME, SystemClock.elapsedRealtime(), AlarmManager.INTERVAL_FIFTEEN_MINUTES, pendingIntent);
        Log.d("widget", "alert on");

    }

    @Override
    public void onDisabled(Context context) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, AppWidget.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, intent, 0);
        alarmManager.cancel(pendingIntent);//알람 해제
        pendingIntent.cancel(); //인텐트 해제
        Log.d("widget", "alert off");
    }


    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        try {
            appDatabase.DatabaseDao().delete(appWidgetIds[0]);
        } catch (Exception e) {
            e.printStackTrace();
            if (appDatabase == null) {
                appDatabase = Room.databaseBuilder(context, AppDatabase.class, "ItemData")
                        .allowMainThreadQueries()
                        .build();

                appDatabase.DatabaseDao().delete(appWidgetIds[0]);
            }
        }

        Log.d("widget", "onDeleted done");
    }

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        String type = null;
        try {

            type = appDatabase.DatabaseDao().getType(appWidgetId);
        } catch (Exception e) {
            e.printStackTrace();
            if (appDatabase == null) {
                appDatabase = Room.databaseBuilder(context, AppDatabase.class, "ItemData")
                        .allowMainThreadQueries()
                        .build();

                type = appDatabase.DatabaseDao().getType(appWidgetId);
            }
        }

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.app_widget);

        Intent intentR = new Intent(context, AppWidget.class);
        intentR.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, intentR, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.refrash, pendingIntent);
        if (type != null) {
            switch (type) {
                case "embedYear":

                    views.setTextViewText(R.id.summery, DEFAULT_YEAR.setTimeItem().getSummery());
                    views.setTextViewText(R.id.percent, DEFAULT_YEAR.setTimeItem().getPercentString() + "%");
                    views.setProgressBar(R.id.progress, 100, (int) Float.parseFloat(DEFAULT_YEAR.setTimeItem().getPercentString()), false);

                    appWidgetManager.updateAppWidget(appWidgetId, views);
                    break;

                case "embedMonth":

                    views.setTextViewText(R.id.summery, DEFAULT_DAY.setTimeItem().getSummery());
                    views.setTextViewText(R.id.percent, DEFAULT_DAY.setTimeItem().getPercentString() + "%");
                    views.setProgressBar(R.id.progress, 100, (int) Float.parseFloat(DEFAULT_DAY.setTimeItem().getPercentString()), false);

                    appWidgetManager.updateAppWidget(appWidgetId, views);
                    break;

                case "embedTime":

                    views.setTextViewText(R.id.summery, DEFAULT_TIME.setTimeItem().getSummery());
                    views.setTextViewText(R.id.percent, DEFAULT_TIME.setTimeItem().getPercentString() + "%");
                    views.setProgressBar(R.id.progress, 100, (int) Float.parseFloat(DEFAULT_TIME.setTimeItem().getPercentString()), false);

                    appWidgetManager.updateAppWidget(appWidgetId, views);
                    break;

                default:
                    AdapterItem adapterItem;

                    //widgetID를 통해 TypeID 검색후 getSelectItem 쿼리를 통해 해당 아이템 객체 불러옴
                    EntityItemInfo entityItemInfo = appDatabase.DatabaseDao().getSelectItem(appDatabase.DatabaseDao().getTypeID(appWidgetId));
                    if (type.equals("Time")) {
                        adapterItem = itemGenerator.generateTimeItem(entityItemInfo);
                    } else {
                        adapterItem = itemGenerator.generateMonthItem(entityItemInfo);
                    }
                    views.setTextViewText(R.id.summery, adapterItem.getSummery());
                    views.setTextViewText(R.id.percent, adapterItem.getPercentString() + "%");
                    views.setProgressBar(R.id.progress, 100, (int) Float.parseFloat(adapterItem.getPercentString()), false);
                    appWidgetManager.updateAppWidget(appWidgetId, views);

            }
        }

        Log.d("widget", type + "update done");


    }

}

File: app/src/main/java/com/devidea/timeleft/AppWidget.kt
package com.devidea.timeleft

import android.appwidget.AppWidgetProvider
import android.content.Intent
import android.appwidget.AppWidgetManager
import android.app.PendingIntent
import android.app.AlarmManager
import android.widget.RemoteViews
import android.content.Context
import android.os.SystemClock
import android.util.Log

class AppWidget : AppWidgetProvider() {

    private val appDatabase = AppDatabase.getInstance(App.context())

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        val action = intent.action
        val appWidgetIds: IntArray
        Log.d("widget", "onReceive() action = $action")
        if (AppWidgetManager.ACTION_APPWIDGET_UPDATE == action) {

            appWidgetIds = appDatabase!!.DatabaseDao().get()!!
            if (appWidgetIds.isNotEmpty()) {
                onUpdate(context, AppWidgetManager.getInstance(context), appWidgetIds)
            }
        }
    }

    override fun onUpdate(
            context: Context,
            appWidgetManager: AppWidgetManager,
            appWidgetIds: IntArray
    ) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
        // There may be multiple widgets active, so update all of them
        for (appWidgetId in appWidgetIds) {
            Log.d("widget", "appWidgetId is $appWidgetId")
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
        val intent = Intent(context, AppWidget::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)

        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

        alarmManager.setInexactRepeating(
                AlarmManager.ELAPSED_REALTIME,
                SystemClock.elapsedRealtime(),
                AlarmManager.INTERVAL_FIFTEEN_MINUTES,
                pendingIntent
        )
        Log.d("widget", "alert on")
    }

    override fun onDisabled(context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, AppWidget::class.java)

        val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)

        alarmManager.cancel(pendingIntent) //알람 해제
        pendingIntent.cancel() //인텐트 해제
        Log.d("widget", "alert off")
    }

    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        super.onDeleted(context, appWidgetIds)
        appDatabase!!.DatabaseDao().delete(appWidgetIds[0])
        Log.d("widget", "onDeleted done")
    }

    private fun updateAppWidget(
            context: Context,
            appWidgetManager: AppWidgetManager,
            appWidgetId: Int
    ) {
        var type: String? = null

        type = appDatabase!!.DatabaseDao().getType(appWidgetId)


        val views = RemoteViews(context.packageName, R.layout.app_widget)
        val intentR = Intent(context, AppWidget::class.java)
        intentR.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE

        val pendingIntent = PendingIntent.getBroadcast(context, 0, intentR, PendingIntent.FLAG_IMMUTABLE)

        views.setOnClickPendingIntent(R.id.refrash, pendingIntent)
        val appIntent = Intent(context, MainActivity::class.java)

        val pe = PendingIntent.getActivity(context, 0, appIntent, PendingIntent.FLAG_IMMUTABLE)
        views.setOnClickPendingIntent(R.id.percent, pe)

        if (type != null) {
            when (type) {
                "embedYear" -> {
                    views.setTextViewText(
                            R.id.summery,
                            MainActivity.ITEM_GENERATE.yearItem().summery
                    )
                    views.setTextViewText(
                            R.id.percent,
                            MainActivity.ITEM_GENERATE.yearItem().percentString + "%"
                    )
                    views.setProgressBar(
                            R.id.progress,
                            100,
                            MainActivity.ITEM_GENERATE.yearItem().percentString
                            !!.toFloat().toInt(),
                            false
                    )
                    appWidgetManager.updateAppWidget(appWidgetId, views)
                }
                "embedMonth" -> {
                    views.setTextViewText(
                            R.id.summery,
                            MainActivity.ITEM_GENERATE.monthItem().summery
                    )
                    views.setTextViewText(
                            R.id.percent,
                            MainActivity.ITEM_GENERATE.monthItem()
                                    .percentString + "%"
                    )
                    views.setProgressBar(
                            R.id.progress,
                            100,
                            MainActivity.ITEM_GENERATE.monthItem().percentString
                            !!.toFloat().toInt(),
                            false
                    )
                    appWidgetManager.updateAppWidget(appWidgetId, views)
                }
                "embedTime" -> {
                    views.setTextViewText(
                            R.id.summery,
                            MainActivity.ITEM_GENERATE.timeItem().summery
                    )
                    views.setTextViewText(
                            R.id.percent,
                            MainActivity.ITEM_GENERATE.timeItem().percentString + "%"
                    )
                    views.setProgressBar(
                            R.id.progress,
                            100,
                            MainActivity.ITEM_GENERATE.timeItem().percentString
                            !!.toFloat().toInt(),
                            false
                    )
                    appWidgetManager.updateAppWidget(appWidgetId, views)
                }
                else -> {
                    val adapterItem: AdapterItem

                    //widgetID를 통해 TypeID 검색후 getSelectItem 쿼리를 통해 해당 아이템 객체 불러옴
                    val entityItemInfo: EntityItemInfo =
                            appDatabase.DatabaseDao().getSelectItem(
                                    appDatabase.DatabaseDao()
                                            .getTypeID(appWidgetId)
                            )
                    adapterItem = if (type == "Time") {
                        MainActivity.ITEM_GENERATE.customTimeItem(entityItemInfo)
                    } else {
                        MainActivity.ITEM_GENERATE.customMonthItem(entityItemInfo)
                    }
                    views.setTextViewText(R.id.summery, adapterItem.summery)
                    views.setTextViewText(R.id.percent, adapterItem.percentString + "%")
                    views.setProgressBar(
                            R.id.progress,
                            100,
                            adapterItem.percentString!!.toFloat().toInt(),
                            false
                    )

                    appWidgetManager.updateAppWidget(appWidgetId, views)
                }

            }
        }
        //Log.d("widget", type + "update done")
    }


}